File: RB101_Programming_Foundations/Small Probems/Easy7/multiply_all_pairs.rb
# input : two arrays
# output : an array whose elements are all possible product permutations of the two input arrays, in sorted ascending order

# first iteratore through the first array, for each element, multiply it by each element in the second array, and store it in a product array

# sort the product array


def multiply_all_pairs(array1, array2)
  product_array = []

  array1.each do |num_array1|
    array2.each do |num_array2|
      product_array << num_array1 * num_array2
    end
  end

  product_array.sort

end


p multiply_all_pairs([2, 4], [4, 3, 1, 2])  == [2, 4, 4, 6, 8, 8, 12, 16]

File: RB101_Programming_Foundations/Small Probems/Easy7/Python/multiply_all_pairs.py
# Write a method that takes two Array arguments in which each Array contains a list of numbers, and returns a new Array that contains the product of every pair of numbers that can be formed between the elements of the two Arrays. The results should be sorted by increasing value.

# You may assume that neither argument is an empty Array.

'''
Input: two arrays

Output: matrix product of two arrays

Rules


Data structure / Algorithms
    create an empty list of length list1*list2
    have an index counter
    iterate through the first list
        for each list1 element, iterate through list2
            multiply list1 element to list2 element, add it to reseult
            increment the index by 1



'''

def multiply_all_pairs(list1, list2):
    result = [0] * ( len(list1) * len(list2) )
    index = 0

    for num1 in list1:
        for num2 in list2:
            result[index] = num1 * num2
            index += 1
    result.sort()
    return result

print(multiply_all_pairs([2, 4], [4, 3, 1, 2]) == [2, 4, 4, 6, 8, 8, 12, 16])
